### ISSUE TITLE
Merge VertexProperty and Property interfaces for implementation consistency

### ISSUE THREAD

**Issue by davidcrouch** [tinkerpop/tinkerpop3#450]
I believe extensibility would be improved if the Property interface included all the methods that VertexProperty does i.e. one Property interface for all. The underlying implementation would return a "not supported" exception for methods where functionality is not provided e.g. meta-properties on edges. 

The benefit of doing this is to allow core graph functionality to be more easily enhanced by graph strategies. For example, by a strategy that supports meta-properties on edges and hides the underlying implementation (a vertex with meta-properties could be used in the raw graph to represent an edge). 

This also opens the door for supporting properties on any element including properties on properties on properties.... etc. (I believe this is similar to RDF). For example, a graph strategy that presents edges as properties on a vertex could easily be accomplished as well as implementing hyper-edges.

**Comment by okram** [comment]
It would be great to remove the distinction between `Property` and `VertexProperty`. If we simply called `VertexProperty` "Property` (i.e. merged their APIs), then we would need to make it such that Properties returned by `Edge` or "VertexProperty" would do the following:

  * `Property.Iterators.propertyIterator()` would return `Iterators.empty()`.
  * `Property.label()` would simply return `Property.key()` as it does now in `VertexProperty` anyways.
  * `Property.property(key,value)` would throw `UnsupportedOperationException`
  * `Property.property(key)` would return `Property.empty()`
  * `Property.keys()` would return `Collections.emptySet()`
  * `Property.value(key)` would trow `UnsupportedOperationException` (or simply say there is no value for that key). 

This was just my initial realization when looking at the API. I'm wondering how this would propagate and effect things such as `instanceof`-type calls and `GraphTraversal`-typing in the core body of code.

@spmallette @mbroecheler

**Comment by davidcrouch** [comment]
Marko, I'm glad it is of interest. I was also thinking renaming VertexProperty to Property was the most likely progression but most importantly because it inherits from Element. That is, three concepts in a graph:- Vertex, Edge, and Property, which all inherit from Element..... and properties could be attached to anything...... which asks why have graph variables?

Now, what if Element included a method like "related" or "associated", which returns all the elements linked to the source element. Could this open up the possibility of a graph vendor then exposing things like hyper-edges through a standard Tinkerpop interface???  Maybe.    Additionally, perhaps a developer could provide this functionality via a Graph Strategy that is used to enhance a graph's functionality. And what if Graph also inherits from Element?  That is, "related" could return other Graph instances and hence support a different form of graph partitioning i.e. a virtual graph built from "relating" graph instances via a federated architecture (obviously up to the graph vendor on how to implement but the Tinkerpop API could support it).

These are just ideas for thought. I think Tinkerpop's role is to provide a concise but flexible standard interface that encourages a graph vendor's creativity rather than hinder it. Knowing that you're getting close to GA just merging/renaming VertexProperty would be huge. Thx.

**Comment by spmallette** [comment]
I think these are all good thoughts, however I feel like I remember having arguments before for just having "Property" and they have ended with us having `VertexProperty` and `Property`.  I don't remember the reasons for the decision, but I believe we exhausted thinking in this area and ended up with what we have.  I have little reason to believe we'll come up with anything new in revisiting this right now.

However, even if we came up with compelling reasons to change it, I believe that we should not.  We are trying to stabilize for GA and this will tear down too much of what has already been done.  Let's leave this for some future change when we believe an API change is prudent and acceptable.

**Comment by okram** [comment]
I concur. The thing I liked about @davidcrouch  argument was the ability to allow vendors that wanted "meta properties" on edges to exist. However, I don't think we should provide such flexibility in the API, when it comes to thinking about how this will work on OLAP engines, Gremlin traversals, etc. I like how we have a "hard coded data model." Moreover, to the RDF nod, the vertexproperty model is analogous to RDF --- save that it provides key/values pairs on all relations.

**Comment by davidcrouch** [comment]
Marko, Stephen, 

I have to agree with you that this request is not feasible for TP 3.0. But I think the request should be kept open due to its nature and many differing opinions.  I found these relevant discussions on the users group:-

https://groups.google.com/forum/#!searchin/gremlin-users/meta$20properties/gremlin-users/4xF_4sKF1tY/Yab9Jqb16_UJ

https://groups.google.com/forum/#!searchin/gremlin-users/meta$20properties/gremlin-users/KqBv3jlj6xM/_cv0UHawvYQJ

https://groups.google.com/forum/#!searchin/gremlin-users/hyperedges/gremlin-users/zRqS8gDhRvk/LJMS3UnXZL8J
 
What is apparent in these great discussions (I think) is the mixing of implementation with the definition of the interfaces, which blurs the arguments made. But I also think the most appropriate direction will become clearer with the adoption and feedback of TP3. 

Besides vendors and end users, I also think there is another audience :- the graph strategy developers (for which I am) who want to develop horizontal abstraction layers on top of the underlying data model of the database. I don't think the Tinkerpop standard/interface should hinder this creativity if the underlying database can support it (either out-of-the-box or with workarounds).

While I understand and somewhat conceptually agree with the arguments against meta-properties on edge properties, I am currently writing a graph strategy through need that contradicts this but am hitting some obstacles, the main one being able to represent them in Gremlin e.g. g.V.outE(has_project).properties("creator").property("acl", "destroy").    I also want to do things like this (hyper-edges):- .g.V.outE(has_project).out("creator").has("name", "Marko").   (Yes, I know how to do this in the current version by defining a different data model). Now, in TP2 we represented properties as separate vertices and values as separate vertices, which gave use the flexibility but the Gremlin traversals were insane.  

I proposed merging Property and VertexProperty because I thought it provided a path to potential solutions for my use cases. If I am missing some other solutions with standard TP3 then i would love to hear about them. But being able to develop custom DSLs is key to being more expressive, which seems to be better promoted with a consistent property interface. However, I have to admit, I have zero experience on the OLAP side and hence cannot comprehend the problems that could arise out of one property concept.

Lastly, great job on TP3. I love the improvements !!!
